### ISSUE TITLE
Caching for Github Actions

### ISSUE THREAD

**Issue by mkaze** [RasaHQ/whatlies#109]
@koaning I was wondering if you have considered using caching in Github Actions ([1][1], [2][2], [3][3]).

Let me give you some stats: currently, downloading the dependency packages with `pip` takes a bit less than 1 minute for Ubuntu and 1.5 minutes for MacOS. Adding torch and its dependencies increases this to 1.5 minutes for Ubuntu (not much increase for MacOS, since its torch package is relatively small). This amount accounts for around one forth of the total test time per workflow. So I am not sure if it's worth it or not to use caching for downloading dependencies (or even test models!). Github gives 5 GB of total cache limit and unlimited number of cache per repository. Though, I am not sure about other pricing and usage factors of Github Action for this repository (since it depends on the type of repository which is not clear to me).

Anyways, I just wanted to bring this up for consideration as well, especially if you are not aware of it.

[1]: https://docs.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows
[2]: https://docs.github.com/en/actions/language-and-framework-guides/using-python-with-github-actions#caching-dependencies
[3]: https://github.com/marketplace/actions/cache

**Comment by sara-tagger** [comment]
Thanks for raising this issue, @akelad will get back to you about it soon✨

###### Please also check out the [docs](https://rasa.com/docs/) and the [forum](https://forum.rasa.com/) in case your issue was raised there too 🤗

**Comment by koaning** [comment]
I've used something similar in gitlab before but haven't used it over on the github side yet. It sounds like a useful feature but there is a catch. 

Pricing is an issue but you typically also want to get a signal if a new version of a dependency is breaking things. For example, if we use a cached version of spaCy 2.3 then we might be blind to breaking changes that spaCy 3.0 might bring. It might be interesting to do a combination of a cron-job that checks the most recent version and a cache that we manually update once in a while.

**Comment by mkaze** [comment]
> For example, if we use a cached version of spaCy 2.3 then we might be blind to breaking changes that spaCy 3.0 might bring

That's not true. If a newer version (compared to the version in the cache) is available on PyPI, then that would be downloaded, installed and then cached for the next time.

As for the pricing, I am not sure. It says that each repo has a 5GB of cache for free, but I don't know if this is counted towards the storage or not.

**Comment by koaning** [comment]
Are you sure? The [docs suggest](https://docs.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows#example-using-the-cache-action) that we might need to manually keep track of versions. In our case that would be ` setup.py` file. But if the file lists `spacy>=2.3.0` then users might download version 3.0.0 while the cache only has 2.3. 

It might still be worth it, but we would need to manually keep updating the `setup.py` file manually to ensure that we've got the most recent version working.

**Comment by mkaze** [comment]
As far as I know, that's the way `pip` works: if a package is not installed first it checks the PyPI to get the info about the newest version, then checks the cache to see if the newest version exists there; if it's in the cache, that one will be used, otherwise it would download the new version from PyPI (e.g. see [this](https://stackoverflow.com/q/53638494/2099607)).

I think the purpose of including requirements file's hash in the cache key is to make sure that a new cache storage would be created when the requirements change, so that it could be used the next time; otherwise, if the cache key is just say `my-pip-cache`, the cache would not include the changed packages and every time those subset of missing packages would be downloaded. In other words, the way I understand it, it's because that cache instances in Github Actions could not be updated; rather, they are just created from scratch or destroyed.
